File: src/core/hebrew_engine.cpp
/**
 * ORKA Hebrew Engine — §7.3: EN ↔ HE (Hebrew + BiDi)
 *
 * Standard QWERTY → Hebrew keyboard mapping.
 * Per §7.3.1:
 *   - EN → HE: Nikud not generated (not on QWERTY)
 *   - HE → EN (no Nikud): standard reverse mapping
 *   - HE → EN (with Nikud): strip Nikud, set nikudStripped=true
 *     so UI can show mandatory warning before action
 *
 * BiDi handling: Raw Bidi markers (U+202B/U+202C) are NOT inserted
 * into fields without Bidi support (per §10.3 test case).
 */

#include "hebrew_engine.h"
#include <unordered_map>
#include <cwctype>

namespace orka {

namespace {

// ── Standard QWERTY → Hebrew mapping ──────────────────────────────
const std::unordered_map<wchar_t, wchar_t> kEnToHe = {
    {L't', L'\u05D0'}, // א (alef)
    {L'c', L'\u05D1'}, // ב (bet)
    {L'd', L'\u05D2'}, // ג (gimel)
    {L's', L'\u05D3'}, // ד (dalet)
    {L'v', L'\u05D4'}, // ה (he)
    {L'u', L'\u05D5'}, // ו (vav)
    {L'z', L'\u05D6'}, // ז (zayin)
    {L'j', L'\u05D7'}, // ח (het)
    {L'y', L'\u05D8'}, // ט (tet)
    {L'h', L'\u05D9'}, // י (yod)
    {L'l', L'\u05DA'}, // ך (final kaf)
    {L'f', L'\u05DB'}, // כ (kaf)
    {L'k', L'\u05DC'}, // ל (lamed)
    {L'o', L'\u05DD'}, // ם (final mem)
    {L'n', L'\u05DE'}, // מ (mem)
    {L'i', L'\u05DF'}, // ן (final nun)
    {L'b', L'\u05E0'}, // נ (nun)
    {L'x', L'\u05E1'}, // ס (samekh)
    {L'g', L'\u05E2'}, // ע (ayin)
    {L';', L'\u05E3'}, // ף (final pe)
    {L'p', L'\u05E4'}, // פ (pe)
    {L'.', L'\u05E5'}, // ץ (final tsadi)
    {L'm', L'\u05E6'}, // צ (tsadi)
    {L'e', L'\u05E7'}, // ק (qof)
    {L'r', L'\u05E8'}, // ר (resh)
    {L'a', L'\u05E9'}, // ש (shin)
    {L',', L'\u05EA'}, // ת (tav)
    // Punctuation pass-through handled separately
    {L'/', L'.'}, 
    {L'w', L','},
    {L'\'',L','},
};

// Build reverse map HE → EN
std::unordered_map<wchar_t, wchar_t> buildReverse() {
    std::unordered_map<wchar_t, wchar_t> rev;
    for (auto& [en, he] : kEnToHe) {
        // Only map Hebrew letters (avoid overwriting with punctuation mappings)
        if (he >= 0x05D0 && he <= 0x05EA) {
            rev[he] = en;
        }
    }
    return rev;
}
const std::unordered_map<wchar_t, wchar_t> kHeToEn = buildReverse();

bool isHebrewLetter(wchar_t ch) {
    return ch >= 0x05D0 && ch <= 0x05EA;
}

bool isNikud(wchar_t ch) {
    return ch >= 0x05B0 && ch <= 0x05C7;
}

bool isLatinLetter(wchar_t ch) {
    return (ch >= L'a' && ch <= L'z') || (ch >= L'A' && ch <= L'Z');
}

} // anonymous namespace


// ════════════════════════════════════════════════════════════════════

Direction HebrewEngine::detectDirection(const std::wstring& input) {
    int latin = 0, hebrew = 0;
    for (wchar_t ch : input) {
        if (isLatinLetter(ch))  ++latin;
        if (isHebrewLetter(ch)) ++hebrew;
    }
    if (latin > hebrew)  return Direction::FORWARD;
    if (hebrew > latin)  return Direction::REVERSE;
    return Direction::UNKNOWN;
}

bool HebrewEngine::containsNikud(const std::wstring& input) {
    for (wchar_t ch : input) {
        if (isNikud(ch)) return true;
    }
    return false;
}

std::wstring HebrewEngine::stripNikud(const std::wstring& input) {
    std::wstring result;
    result.reserve(input.size());
    for (wchar_t ch : input) {
        if (!isNikud(ch)) result.push_back(ch);
    }
    return result;
}

ConversionResult HebrewEngine::convert(const std::wstring& input) {
    ConversionResult result;
    result.nikudStripped = false;
    result.escapeRequired = false;

    Direction dir = detectDirection(input);

    if (dir == Direction::FORWARD || dir == Direction::UNKNOWN) {
        // EN → HE
        std::wstring out;
        out.reserve(input.size());
        for (wchar_t ch : input) {
            wchar_t lower = std::towlower(ch);
            auto it = kEnToHe.find(lower);
            if (it != kEnToHe.end()) {
                out.push_back(it->second);
            } else {
                // Pass-through: digits, spaces, punctuation not in map
                out.push_back(ch);
            }
        }
        result.text = out;
        result.success = true;
    } else {
        // HE → EN
        // §7.3.1: Check for Nikud — lossy conversion warning
        if (containsNikud(input)) {
            result.nikudStripped = true;
            // Strip Nikud before conversion
        }

        std::wstring cleaned = result.nikudStripped ? stripNikud(input) : input;
        std::wstring out;
        out.reserve(cleaned.size());

        for (wchar_t ch : cleaned) {
            auto it = kHeToEn.find(ch);
            if (it != kHeToEn.end()) {
                out.push_back(it->second);
            } else {
                out.push_back(ch);
            }
        }
        result.text = out;
        result.success = true;
    }

    return result;
}

} // namespace orka

File: src/hebrew_engine.rs
// ═══════════════════════════════════════════════════════
// Orka — Hebrew Engine (EN ↔ HE)
//
// Standard QWERTY → Hebrew keyboard mapping.
// Handles Nikud (U+05B0–U+05C7) lossy stripping with warning.
//
// Ported from C++ hebrew_engine.cpp
// ═══════════════════════════════════════════════════════

use std::collections::HashMap;
use crate::conversion_engine::ConversionResult;
use crate::hangul_engine::Direction;

pub struct HebrewEngine;

impl HebrewEngine {
    /// Detect direction: Latin → Forward, Hebrew → Reverse
    pub fn detect_direction(input: &str) -> Direction {
        let mut latin = 0usize;
        let mut hebrew = 0usize;
        for ch in input.chars() {
            if ch.is_ascii_alphabetic() { latin += 1; }
            if is_hebrew_letter(ch) { hebrew += 1; }
        }
        if latin > hebrew { Direction::Forward }
        else if hebrew > latin { Direction::Reverse }
        else { Direction::Unknown }
    }

    /// Check if text contains Nikud diacritics
    pub fn contains_nikud(input: &str) -> bool {
        input.chars().any(is_nikud)
    }

    /// Strip Nikud from Hebrew text
    pub fn strip_nikud(input: &str) -> String {
        input.chars().filter(|c| !is_nikud(*c)).collect()
    }

    /// Convert EN ↔ HE text
    pub fn convert(input: &str) -> ConversionResult {
        let en_to_he = build_en_to_he();
        let he_to_en = build_he_to_en(&en_to_he);
        let dir = Self::detect_direction(input);

        if dir == Direction::Forward || dir == Direction::Unknown {
            // EN → HE
            let out: String = input.chars().map(|ch| {
                let lower = ch.to_lowercase().next().unwrap_or(ch);
                *en_to_he.get(&lower).unwrap_or(&ch)
            }).collect();

            ConversionResult {
                text: out, success: true,
                nikud_stripped: false, escape_required: false, changed_count: 0,
            }
        } else {
            // HE → EN
            let nikud_stripped = Self::contains_nikud(input);
            let cleaned = if nikud_stripped { Self::strip_nikud(input) } else { input.to_string() };

            let out: String = cleaned.chars().map(|ch| {
                *he_to_en.get(&ch).unwrap_or(&ch)
            }).collect();

            ConversionResult {
                text: out, success: true,
                nikud_stripped, escape_required: false, changed_count: 0,
            }
        }
    }
}

// ── Helpers ────────────────────────────────────────────

fn is_hebrew_letter(ch: char) -> bool {
    let cp = ch as u32;
    cp >= 0x05D0 && cp <= 0x05EA
}

fn is_nikud(ch: char) -> bool {
    let cp = ch as u32;
    cp >= 0x05B0 && cp <= 0x05C7
}

// ── Standard QWERTY → Hebrew keyboard mapping ────────
fn build_en_to_he() -> HashMap<char, char> {
    HashMap::from([
        ('t', '\u{05D0}'), // א (alef)
        ('c', '\u{05D1}'), // ב (bet)
        ('d', '\u{05D2}'), // ג (gimel)
        ('s', '\u{05D3}'), // ד (dalet)
        ('v', '\u{05D4}'), // ה (he)
        ('u', '\u{05D5}'), // ו (vav)
        ('z', '\u{05D6}'), // ז (zayin)
        ('j', '\u{05D7}'), // ח (het)
        ('y', '\u{05D8}'), // ט (tet)
        ('h', '\u{05D9}'), // י (yod)
        ('l', '\u{05DA}'), // ך (final kaf)
        ('f', '\u{05DB}'), // כ (kaf)
        ('k', '\u{05DC}'), // ל (lamed)
        ('o', '\u{05DD}'), // ם (final mem)
        ('n', '\u{05DE}'), // מ (mem)
        ('i', '\u{05DF}'), // ן (final nun)
        ('b', '\u{05E0}'), // נ (nun)
        ('x', '\u{05E1}'), // ס (samekh)
        ('g', '\u{05E2}'), // ע (ayin)
        (';', '\u{05E3}'), // ף (final pe)
        ('p', '\u{05E4}'), // פ (pe)
        ('.', '\u{05E5}'), // ץ (final tsadi)
        ('m', '\u{05E6}'), // צ (tsadi)
        ('e', '\u{05E7}'), // ק (qof)
        ('r', '\u{05E8}'), // ר (resh)
        ('a', '\u{05E9}'), // ש (shin)
        (',', '\u{05EA}'), // ת (tav)
        ('/', '.'),
        ('w', ','),
    ])
}

fn build_he_to_en(en_to_he: &HashMap<char, char>) -> HashMap<char, char> {
    let mut rev = HashMap::new();
    for (&en, &he) in en_to_he {
        // Only map Hebrew letters (avoid overwriting with punctuation mappings)
        let cp = he as u32;
        if cp >= 0x05D0 && cp <= 0x05EA {
            rev.insert(he, en);
        }
    }
    rev
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_direction_latin() {
        assert_eq!(HebrewEngine::detect_direction("hello"), Direction::Forward);
    }

    #[test]
    fn test_direction_hebrew() {
        assert_eq!(HebrewEngine::detect_direction("\u{05E9}\u{05DC}\u{05D5}\u{05DD}"), Direction::Reverse);
    }

    #[test]
    fn test_en_to_he_single() {
        let res = HebrewEngine::convert("a");
        assert!(res.success);
        assert_eq!(res.text, "\u{05E9}"); // ש
    }

    #[test]
    fn test_en_to_he_alef() {
        let res = HebrewEngine::convert("t");
        assert_eq!(res.text, "\u{05D0}"); // א
    }

    #[test]
    fn test_en_to_he_word() {
        let res = HebrewEngine::convert("akln");
        assert!(res.success);
        assert_eq!(res.text, "\u{05E9}\u{05DC}\u{05DA}\u{05DE}"); // שלךמ
    }

    #[test]
    fn test_he_to_en_single() {
        let res = HebrewEngine::convert("\u{05E9}");
        assert!(res.success);
        assert!(!res.nikud_stripped);
        assert_eq!(res.text, "a");
    }

    #[test]
    fn test_nikud_detection() {
        assert!(HebrewEngine::contains_nikud("\u{05E9}\u{05B8}"));
        assert!(!HebrewEngine::contains_nikud("\u{05E9}\u{05DC}"));
    }

    #[test]
    fn test_nikud_stripping() {
        let stripped = HebrewEngine::strip_nikud("\u{05E9}\u{05B8}\u{05DC}");
        assert_eq!(stripped, "\u{05E9}\u{05DC}");
    }

    #[test]
    fn test_nikud_lossy_warning() {
        let with_nikud = "\u{05E9}\u{05B8}\u{05C1}\u{05DC}\u{05D5}\u{05B9}\u{05DD}";
        let res = HebrewEngine::convert(with_nikud);
        assert!(res.nikud_stripped);
        assert!(res.success);
    }

    #[test]
    fn test_roundtrip() {
        let fwd = HebrewEngine::convert("test");
        let rev = HebrewEngine::convert(&fwd.text);
        assert_eq!(rev.text, "test");
    }

    #[test]
    fn test_case_insensitive() {
        let lower = HebrewEngine::convert("a");
        let upper = HebrewEngine::convert("A");
        assert_eq!(lower.text, upper.text);
    }
}
